documentation/projects/proposals/popularity_optimizations/20230406-project_proposal_popularity_optimizations.md

```diff
@@ -0,0 +1,120 @@
+acceptable to recalculate constants and scores on a quarterly basis.
+
+## Success
+
+<!-- How do we measure the success of the project? How do we know our ideas worked? -->
+
+For this project to be considered a success, we must be able to turn on the data
+refresh DAGs with an automated weekly schedule. This requires that the data
+refresh must be able to complete in under a week.
+
+The requirements in the section above should also be met to prevent regressions
+in data quality and availability.
+
+## Participants and stakeholders
+
+<!-- Who is working on the project and who are the external stakeholders, if any? Consider the lead, implementers, designers, and other stakeholders who have a say in how the project goes. -->
+
+- Lead: @stacimc
+- Implementation:
+  - @stacimc
+  - TBD
+- Stakeholders:
+  - Openverse Team
+
+## Infrastructure
+
+<!-- What infrastructural considerations need to be made for this project? If there are none, say so explicitly rather than deleting the section. -->
+
+This project will require:
+
+- changes to the Catalog database schema
+- creation and modification of Airflow DAGs related to the data refresh
+- minor changes in the ingestion server
+
+Specifics will be detailed in the implementation plan. For testing, this will
+require connecting Airflow with the staging ingestion server and API database.
+
+I do not anticipate changes to the Elasticsearch mapping or queries. However it
+should be noted that if this changes, we may require access to the Search
+Relevancy Sandbox in order to test.
+
+## Accessibility
+
+<!-- Are there specific accessibility concerns relevant to this project? Do you expect new UI elements that would need particular care to ensure they're implemented in an accessible way? Consider also low-spec device and slow internet accessibility, if relevant. -->
+
+This project should not have any frontend facing changes. However it will touch
+many complex parts of ingestion, popularity calculation, and the data refresh,
+and there must be a significant effort at each stage to document both previously
+undocumented processes and the additions made here.
+
+## Marketing
+
+<!-- Are there potential marketing opportunities that we'd need to coordinate with the community to accomplish? If there are none, say so explicitly rather than deleting the section. -->
+
+This change is largely internal and may present few marketing opportunities.
+However, since it will enable us to get automated data refreshes running
+consistently, there may be a significant increase in records in Elasticsearch
+after the project completes. We should check the record count difference and
+make an announcement if it is significant.
+
+## Required implementation plans
+
+<!-- What are the required implementation plans? Consider if they should be split per level of the stack or per feature. -->
+
+1. Removing the popularity steps from the data refresh DAG
```

Comment:
I have the same question as @obulat. Will this be the only plan required? If so, I think the project name should be adjusted to something like "Reenabling Image Data Refresh". The requirements are more focused on the whole process, and otherwise, it's a bit confusing.

Reply:
The idea is to separate the popularity steps from the data refresh DAG, such that they can be run independently. The _effect_ is that we can re-enable the data refresh (without failing to meet other requirements), but I was trying to capture what the content of the IP will be rather than the purpose of the project, which is the focus of _this_ document.

I do think this will be a single implementation plan, because it cannot be realistically broken into separate pieces. I did not add more details here because as changes are made, the project proposal should not drift from the IP.

Reply:
I got your point. It's perfectly valid as a goal to separate those processes. For example, should we then name this "Decoupling the popularity calculations of the Data Refresh"? The proposal suggests that after the project is complete, the calculations could be moved to a "quarterly basis" but that sounds contradictory if the project is looking to optimize that process 🤔 Shouldn't it be possible for it to run more often after completion?

Reply:
The goal of the project as defined in the Project Summary is to reduce the length of the _data refresh_, which is what we care about running more frequently. The mechanism for doing that is optimizing the popularity calculations, which is the longest running step. That could mean making those steps literally run faster, or it could mean decoupling them from the data refresh. I'm proposing the latter.

I tried to capture in the Requirements list what we need from the optimizations. Basically: the data refresh should be able to run at any time without waiting on a popularity task; data should always have popularity scores before being added to the API, but we're okay with the scores being _updated_ slightly less frequently.

I think possibly it's confusing that the project itself was named "Popularity optimizations" instead of "Data refresh optimizations", but I think the name is still accurate. We are optimizing the way we calculate popularity scores in order to speed up data refreshes.

Reply:
Thank you for this extended comment. I understand that the exact implementation details are not yet fully clear, so having some idea of what they can be is helpful